### DISCUSSION TITLE
Creating instances of opaque types from `.json` file data

### DISCUSSION THREAD

**Discussion by TerensTare** [nlohmann/json#2260] [Q&A] [answered]
Hello there. I am using SDL2 together with nlohmann::json (this library), and I wanted to load some `SDL_Texture`-s from a `.json` file, which contains the names of some image files. The problem is that `SDL_Texture` is an opaque data type (the struct's definition is on a C source(`.c`) file rather than on a header(`.h`) file). However, SDL2 provides some factory functions which need a filename and a renderer pointer (not related to the question). I already read the examples section on `README.md`, but I can't find anything that can help me.

**Comment by nlohmann** [comment, marked as answer]
In order to asses your issue, we need the following information:

- What is the issue you have?

- Please describe the steps to reproduce the issue. Can you provide a small but working code example?

- What is the expected behavior?

- And what is the actual behavior instead?

- Which compiler and operating system are you using? Is it a [supported compiler](https://github.com/nlohmann/json#supported-compilers)?

- Did you use a released version of the library or the version from the `develop` branch?

- If you experience a compilation error: can you [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests)?

**Reply by TerensTare** [reply]
Sorry for not giving enough information.
To start with, I am using `vcpkg`'s latest updated port of `nlohmann::json` with the `x64-windows` triplet. The OS is Windows 10 with the Visual Studio 2019 C++ compiler. I am not having any compilation or even any problem with code not behaving correctly.
The problem I have is that I want to load an image to the screen, and get the name of the image file from a `.json` file.
ex.:
``` json
{
    "width": 100,
    "height": 100,
    "file": "myImage.png"
}
```
 I use SDL2 (written in C) for handling graphics. It provides a `SDL_Texture` structure for handling images/textures. But `SDL_Texture` is an opaque type (meaning that on the header file it is just `typedef struct SDL_Texture SDL_Texture;`, and the body of the structure is defined on a `.c` file), and SDL_Image provides a function named `IMG_LoadTexture` for creating `SDL_Texture*` for you (the user). I can easily parse/extract `width` and `height` from the `.json` file, but I am not sure about the way I should specialise `nlohmann::adl_serializer` for `SDL_Texture`.

**Reply by nlohmann** [reply]
Did you try to define a `to_json`/`from_json` function for `SDL_Texture`?

**Reply by TerensTare** [reply]
As a matter of fact, I only tried to define an specialisation of `nlohmann::adl_serializer<SDL_Texture>` by following the examples section on `README.md`. At first I didn't fully understand how `basic_json` and its members worked.
I used to write sth like:
``` c++
namespace nlohmann {
template<> struct adl_serializer<SDL_Texture> {
static void from_json(nlohmann::json const& j, SDL_Texture* tex) {
j.at("file").get_to(tex): // I thought using get_to was mandatory, that's why I couldn't figure out a solution
}
};
}
```

But after I walked through the whole examples, checked the return types of `basic_json` member functions and experimented a bit with them, I came up with this solution. So basically I wrote a `TextureManager` class. It's only purpose is to handle the lifetime of `SDL_Texture`-s for the user. I added a `from_json` function defined like this:
``` c++
SDL_Texture* TextureManager::from_json(SDL_Renderer* ren, nlohmann::json const& j)
{
char const* file{j.at("file")};
SDL_Texture* ret = IMG_LoadTexture(ren, file);
}
```

Now I use that function to load `SDL_Texture`-s for me. Thank you for everything, your help and the examples were invaluable for me.
